os/kernel/binary_manager/binary_manager_bootparam.c

```diff
@@ -422,3 +425,96 @@ void binary_manager_set_bpidx(uint8_t index)
 {
 	g_bp_info.inuse_idx = index;
 }
+
+void binary_manager_swap_bootparam(int requester_pid)
+{
+	int ret;
+	char *bootparam;
+	bool is_all_updatable;
+	char q_name[BIN_PRIVMQ_LEN];
+	binmgr_bpdata_t update_bp_data;
+	binmgr_response_t response_msg;
+#ifdef CONFIG_APP_BINARY_SEPARATION
+	int bin_idx;
+	uint32_t bin_count;
+#endif
+
+	memset((void *)&response_msg, 0, sizeof(binmgr_response_t));
+
+	if (requester_pid < 0) {
+		bmdbg("Invalid requester pid %d\n", requester_pid);
+		return;
+	}
+
+	bootparam = (char *)kmm_malloc(BOOTPARAM_SIZE);
+	if (!bootparam) {
+		bmdbg("Fail to malloc to read BP\n");
+		ret = BINMGR_OUT_OF_MEMORY;
+		goto send_response;
+	}
+	memset(bootparam, 0xff, BOOTPARAM_SIZE);
+
+	response_msg.result = BINMGR_OK;
+	is_all_updatable = true;
+
+	/* Get current bootparam data */
+	memcpy(&update_bp_data, binary_manager_get_bpdata(), sizeof(binmgr_bpdata_t));
+	update_bp_data.version++;
+
+	/* Update bootparam and Reboot if valid kernel binary exists */
+	ret = binary_manager_check_kernel_update(false);
+	if (ret < 0) {
+		bmdbg("Fail to find valid kernel binary, %d\n", ret);
+		goto send_response;
+	}
+	update_bp_data.active_idx ^= 1;
+
+#ifdef CONFIG_RESOURCE_FS
+	/* Update bootparam if valid resource binary exists */
+	ret = binary_manager_check_resource_update(false);
+	if (ret < 0) {
```

Comment:
When we get failure at somewhere, kernel update or resource update or user update, can we know which gets failure?